uaclient/tests/test_lock.py

```diff
@@ -0,0 +1,219 @@
+M_PATH_UACONFIG = "uaclient.config.UAConfig."
+
+
+@mock.patch("os.getpid", return_value=123)
+@mock.patch(M_PATH_UACONFIG + "delete_cache_key")
+@mock.patch(M_PATH_UACONFIG + "add_notice")
+@mock.patch(M_PATH_UACONFIG + "write_cache")
+class TestSingleAttemptLock:
+    def test_creates_and_releases_lock(
+        self,
+        m_write_cache,
+        m_add_notice,
+        m_delete_cache_key,
+        _m_getpid,
+        FakeConfig,
+    ):
+        cfg = FakeConfig()
+        arg, kwarg = mock.sentinel.arg, mock.sentinel.kwarg
+
+        def test_function(args, cfg):
+            assert arg == mock.sentinel.arg
+            assert kwarg == mock.sentinel.kwarg
+            return mock.sentinel.success
+
+        with SingleAttemptLock(cfg, "some operation"):
+            ret = test_function(arg, cfg=cfg)
+
+        assert mock.sentinel.success == ret
+        assert [
+            mock.call("lock", "123:some operation")
+        ] == m_write_cache.call_args_list
+        lock_msg = "Operation in progress: some operation"
+        assert [mock.call("", lock_msg)] == m_add_notice.call_args_list
+        assert [mock.call("lock")] == m_delete_cache_key.call_args_list
+
+    def test_creates__and_releases_lock_when_error_occurs(
+        self,
+        m_write_cache,
+        m_add_notice,
+        m_delete_cache_key,
+        _m_getpid,
+        FakeConfig,
+    ):
+        cfg = FakeConfig()
+
+        def test_function():
+            raise RuntimeError("test")
+
+        with pytest.raises(RuntimeError) as exc:
+            with SingleAttemptLock(cfg, "some operation"):
+                test_function()
+
+        assert "test" == str(exc.value)
+        assert [
+            mock.call("lock", "123:some operation")
+        ] == m_write_cache.call_args_list
+        lock_msg = "Operation in progress: some operation"
+        assert [mock.call("", lock_msg)] == m_add_notice.call_args_list
+        assert [mock.call("lock")] == m_delete_cache_key.call_args_list
+
+    @mock.patch(M_PATH_UACONFIG + "check_lock_info", return_value=(10, "held"))
+    def test_raises_lock_held_when_held(
+        self,
+        m_write_cache,
+        m_add_notice,
+        m_delete_cache_key,
+        _m_getpid,
+        _m_check_lock_info,
+        FakeConfig,
+    ):
+        cfg = FakeConfig()
+        arg, kwarg = mock.sentinel.arg, mock.sentinel.kwarg
+
+        def test_function(args, cfg):
+            assert arg == mock.sentinel.arg
+            assert kwarg == mock.sentinel.kwarg
+            return mock.sentinel.success
+
+        with pytest.raises(LockHeldError) as exc:
+            with SingleAttemptLock(cfg, "some operation"):
+                test_function(arg, cfg=cfg)
+
+        assert exc.value.lock_holder == "held"
+        assert exc.value.lock_holder_pid == 10
+        assert exc.value.lock_request == "some operation"
+
+        assert [
+            mock.call("lock", "123:some operation")
+        ] != m_write_cache.call_args_list
```

Comment:
Should we also assert that `m_write_cache` `call_args_list` is empty ?

Reply:
Yes that would be a better assertion for sure